applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.applicationconfig;
+
+import com.azure.common.http.HttpClient;
+import com.azure.common.http.HttpPipeline;
+import com.azure.common.http.policy.AsyncCredentialsPolicy;
+import com.azure.common.http.policy.HttpLogDetailLevel;
+import com.azure.common.http.policy.HttpLoggingPolicy;
+import com.azure.common.http.policy.HttpPipelinePolicy;
+import com.azure.common.http.policy.RetryPolicy;
+import com.azure.common.http.policy.UserAgentPolicy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Provides configuration options for instances of {@link ConfigurationClient}.
+ *
+ * @see ConfigurationClient
+ * @see ConfigurationClientCredentials
+ */
+public final class ConfigurationClientBuilder {
+    //TODO: Fix with https://github.com/Azure/azure-sdk-for-java/issues/3141
+    static final String SDK_NAME = "Azure-Application-Configuration";
+    static final String SDK_VERSION = "1.0.0-SNAPSHOT";
+
+    private final List<HttpPipelinePolicy> policies;
+    private ConfigurationClientCredentials credentials;
+    private HttpClient httpClient;
+    private HttpLogDetailLevel httpLogDetailLevel;
+    private RetryPolicy retryPolicy;
+    private String userAgent;
+
+    ConfigurationClientBuilder() {
+        userAgent = String.format("Azure-SDK-For-Java/%s (%s)", SDK_NAME, SDK_VERSION);
+        retryPolicy = new RetryPolicy();
+        httpLogDetailLevel = HttpLogDetailLevel.NONE;
+        policies = new ArrayList<>();
+    }
+
+    /**
+     * Creates a {@link ConfigurationClient} based on options set in the Builder.
+     *
+     * <p>
+     * Every time {@code build()} is called, a new instance of {@link ConfigurationClient} is created.
+     *
+     * @return A ConfigurationClient with the options set from the builder.
+     * @throws IllegalStateException If {@link ConfigurationClientBuilder#credentials(ConfigurationClientCredentials)} has not been set.
+     */
+    public ConfigurationClient build() {
+        if (credentials == null) {
+            throw new IllegalStateException("'credentials' is required.");
+        }
+
+        // Closest to API goes first, closest to wire goes last.
+        final List<HttpPipelinePolicy> policies = new ArrayList<>();
+
+        policies.add(new UserAgentPolicy(userAgent));
+        policies.add(new RequestIdPolicy());
+        policies.add(retryPolicy);
+        policies.add(new ConfigurationCredentialsPolicy());
+        policies.add(new AsyncCredentialsPolicy(credentials));
+
+        policies.addAll(this.policies);
+
+        policies.add(new HttpLoggingPolicy(httpLogDetailLevel));
+
+        HttpPipeline pipeline = httpClient == null
+            ? new HttpPipeline(policies)
+            : new HttpPipeline(httpClient, policies);
```

Comment:
I wasn't sure if we should allow users to pass in an entire pipeline... I guess I'm worried that they could shoot themselves in the foot by not bothering to add in a credential policy and as a result, have something that doesn't work at all.

Reply:
I understand the reluctance. The exact phrase used in the arch board is "falling off the cliff". It's considered an escape hatch that might be useful in certain circumstances. I would recommend you consider making it settable, and make sure to update @g2vinay so he can follow an equivalent pattern in key vault.

Reply:
I see I see. 🤔
A couple of approaches could be to build the ConfigurationClient if they specify the pipeline since we don't need any other options.
```java
class ConfigurationClientBuilder {
    ConfigurationClient pipeline(HttpPipeline);
```
or
```java
class ConfigurationClientBuilder {
    ConfigurationClientBuilder pipeline(HttpPipeline);
```
And if the `pipeline` is set, to ignore all the other options and build the ConfigurationClient with that.

Reply:
For consistency I would go with the later approach, and document the precedence order.